mono/metadata/icall.c

```diff
@@ -3611,117 +3611,135 @@ write_enum_value (char *mem, int type, guint64 value)
 	return;
 }
 
-ICALL_EXPORT MonoObject *
-ves_icall_System_Enum_ToObject (MonoReflectionType *enumType, guint64 value)
+ICALL_EXPORT MonoObjectHandle
+ves_icall_System_Enum_ToObject (MonoReflectionTypeHandle enumType, guint64 value, MonoError *error)
 {
-	ERROR_DECL (error);
 	MonoDomain *domain; 
 	MonoClass *enumc;
-	MonoObject *res;
+	MonoObjectHandle res;
 	MonoType *etype;
+	guint res_gchandle = 0;
 
-	domain = mono_object_domain (enumType); 
-	enumc = mono_class_from_mono_type (enumType->type);
+	domain = MONO_HANDLE_DOMAIN (enumType);
+	enumc = mono_class_from_mono_type (MONO_HANDLE_GETVAL (enumType, type));
 
 	mono_class_init_checked (enumc, error);
-	if (mono_error_set_pending_exception (error))
-		return NULL;
+	goto_if_nok (error, return_null);
 
 	etype = mono_class_enum_basetype (enumc);
 
-	res = mono_object_new_checked (domain, enumc, error);
-	if (mono_error_set_pending_exception (error))
-		return NULL;
-	write_enum_value ((char *)res + sizeof (MonoObject), etype->type, value);
+	res = mono_object_new_handle (domain, enumc, error);
+	goto_if_nok (error, return_null);
+
+	write_enum_value (mono_object_handle_pin_unbox (res, &res_gchandle), etype->type, value);
+	mono_gchandle_free (res_gchandle);
 
 	return res;
+
+return_null:
+	return MONO_HANDLE_NEW (MonoObject, NULL);
 }
 
 ICALL_EXPORT MonoBoolean
-ves_icall_System_Enum_InternalHasFlag (MonoObject *a, MonoObject *b)
+ves_icall_System_Enum_InternalHasFlag (MonoObjectHandle a, MonoObjectHandle b, MonoError *error)
 {
-	int size = mono_class_value_size (a->vtable->klass, NULL);
+	int size = mono_class_value_size (MONO_HANDLE_GETVAL (a, vtable)->klass, NULL);
 	guint64 a_val = 0, b_val = 0;
+	guint a_gchandle = 0, b_gchandle = 0;
+
+	memcpy (&a_val, mono_object_handle_pin_unbox (a, &a_gchandle), size);
```

Comment:
This is killing a fly with a rocket, we're creating/destroying a GC handle to read a single word.

Reply:
@lambdageek This is analogous to reading a field, so we should offer something similar to MONO_HANDLE_GET that does the job of copying non-ref bytes out of a handle. Could you work with Jay on adding such primitive?

Something as simple as this could work:
```c
#define MONO_HANDLE_UNBOX_COPY_TO(HANDLE, DEST, COUNT) do {			\
		MONO_HANDLE_SUPPRESS (memcpy((DEST), mono_object_unbox ((gpointer)MONO_HANDLE_RAW (MONO_HANDLE_UNSUPPRESS (HANDLE)), (COUNT)); \
	} while (0)
```

Reply:
Can fixed help here?
I get wierd results:
```
unsafe class A
{
public enum E { zero, one, two };
public E e = new E();

void f1(E *pe) { }

void F2(E f)
{
	fixed (E * pi = &f)
	{
		f1(pi);
	}
}
};

$ csc 1.cs /unsafe
Microsoft (R) Visual C# Compiler version 2.3.1.61919 (57c81319)
Copyright (C) Microsoft Corporation. All rights reserved.

1.cs(10,18): error CS0213: You cannot use the fixed statement to take the address of an already fixed expression

Ok, how about then:
```unsafe class A
{
public enum E { zero, one, two };
public E e = new E();

void f1(E *pe) { }

void F2(E f)
{
	f1(&f);
}
};```

and then assume enum==int?